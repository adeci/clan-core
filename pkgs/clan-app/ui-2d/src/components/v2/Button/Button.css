.button {
  @apply flex gap-2 shrink-0 items-center justify-center;
  @apply px-4 py-2;

  height: theme(height.9);
  border-radius: 3px;

  /* Add transition for smooth width animation */
  transition: width 0.5s ease 0.1s;

  &.s {
    @apply px-3 py-1.5;
    height: theme(height.7);
    border-radius: 2px;

    &:has(> .icon-start):has(> .label) {
      @apply pl-2;
    }

    &:has(> .icon-end):has(> .label) {
      @apply pr-2;
    }
  }

  &.primary {
    @apply bg-inv-acc-4 fg-inv-1;
    @apply border border-solid border-inv-4;
    @apply shadow-button-primary;

    &.ghost {
      @apply bg-transparent border-none shadow-none;
    }

    &:hover {
      @apply bg-inv-acc-3 border-solid shadow-button-primary-hover;
      border-color: theme(backgroundColor.secondary.700);
    }

    &:active {
      @apply bg-inv-acc-4 border-solid border-inv-3 shadow-button-primary-active;
    }

    &:focus-visible {
      @apply bg-inv-acc-4 border-solid border-inv-3 shadow-button-primary-focus;
    }

    &:disabled {
      @apply bg-def-acc-3 border-solid border-def-3 fg-def-3 shadow-none;
    }

    & > .icon {
      @apply fg-inv-1;
    }
  }

  &.secondary {
    @apply bg-def-acc-2 fg-def-1;
    @apply border border-solid border-inv-2;
    @apply shadow-button-secondary;

    &.ghost {
      @apply bg-transparent border-none shadow-none;
    }

    &:hover {
      @apply bg-def-acc-3 border-solid shadow-button-secondary-hover;
      border-color: theme(backgroundColor.secondary.700);
    }

    &:focus-visible {
      @apply bg-def-acc-3 border-solid border-inv-3 shadow-button-secondary-focus;
    }

    &:active {
      @apply bg-def-acc-3 border-solid border-inv-4 shadow-button-secondary-active;
    }

    &:disabled {
      @apply bg-def-2 border-solid border-def-2 fg-def-3 shadow-none;
    }

    & > .icon {
      @apply fg-def-1;

      &.icon-loading {
        color: #0051ff;
      }
    }
  }

  &.icon {
    @apply p-2;
  }

  &:has(> .icon-start):has(> .label) {
    @apply pl-3.5;
  }

  &:has(> .icon-end):has(> .label) {
    @apply pr-3.5;
  }

  & > div.loader {
    @apply w-0 opacity-0;
    @apply top-0 left-0 -mr-2;
    transition: all 0.5s ease;
  }

  &.loading {
    @apply cursor-wait;

    & > div.loader {
      @apply w-4 opacity-100;
      margin-right: revert;
      transition: all 0.5s ease;
    }
  }
}

/* button group */
.button-group .button:first-child {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group .button:first-child {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group .button:last-child {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}
