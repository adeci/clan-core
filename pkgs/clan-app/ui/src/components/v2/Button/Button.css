.button {
  @apply flex gap-2 shrink-0 items-center justify-center;
  @apply px-4 py-2;

  height: theme(height.9);
  border-radius: 3px;

  /* Add transition for smooth width animation */
  transition: width 0.5s ease 0.1s;

  &.s {
    @apply px-3 py-1.5;
    height: theme(height.7);
    border-radius: 2px;

    &:has(> .icon-start):has(> .label) {
      @apply pl-2;
    }

    &:has(> .icon-end):has(> .label) {
      @apply pr-2;
    }
  }

  &.primary {
    @apply bg-inv-acc-4 fg-inv-1;
    @apply border border-solid border-inv-4;

    box-shadow: 0.125rem 0.125rem 0 0 theme(colors.bg.inv.acc.3) inset;

    &.ghost {
      @apply bg-transparent border-none shadow-none;
    }

    &:hover {
      @apply bg-inv-acc-3 border-solid border-inv-acc-3;
      box-shadow: 0.125rem 0.125rem 0 0 theme(colors.bg.inv.acc.2) inset;
    }

    &:active {
      @apply bg-inv-acc-4 border-solid border-inv-3;
      box-shadow:
        0 0 0 0.0625rem theme(colors.off.white),
        0 0 0 0.125rem theme(colors.bg.inv.acc.4),
        -0.125rem -0.125rem 0 0 theme(colors.bg.inv.acc.1) inset;
    }

    &:focus-visible {
      @apply bg-inv-acc-4 border-solid border-inv-3;
      box-shadow:
        0 0 0 0.0625rem theme(colors.off.white),
        0 0 0 0.125rem theme(colors.border.semantic.info.1),
        0.125rem 0.125rem 0 0 theme(colors.bg.inv.acc.2) inset;
    }

    &:disabled {
      @apply bg-def-acc-3 border-solid border-def-3 fg-def-3 shadow-none;
    }

    & > .icon {
      @apply fg-inv-1;
    }
  }

  &.secondary {
    @apply bg-def-acc-2 fg-def-1;
    @apply border border-solid border-inv-2;

    box-shadow:
      -0.125rem -0.125rem 0 0 #cedfe2 inset,
      0.125rem 0.125rem 0 0 theme(colors.off.white) inset;

    &.ghost {
      @apply bg-transparent border-none shadow-none;
    }

    &:hover {
      @apply bg-def-acc-3 border-solid border-inv-3;
      box-shadow:
        -0.125rem -0.125rem 0 0 #cedfe2 inset,
        0.125rem 0.125rem 0 0 theme(colors.off.white) inset;
    }

    &:focus-visible {
      @apply bg-def-acc-3 border-solid border-inv-3;
      box-shadow:
        0 0 0 0.0625rem theme(colors.off.white),
        0 0 0 0.125rem theme(colors.border.semantic.info.1),
        -0.125rem -0.125rem 0 0 #cedfe2 inset,
        0.125rem 0.125rem 0 0 theme(colors.off.white) inset;
    }

    &:active {
      @apply bg-def-acc-3 border-solid border-inv-4;
      box-shadow:
        0 0 0 0.0625rem theme(colors.off.white),
        0 0 0 0.125rem theme(colors.bg.inv.acc.4),
        0.125rem 0.125rem 0 0 theme(colors.bg.inv.acc.2) inset;
    }

    &:disabled {
      @apply bg-def-2 border-solid border-def-2 fg-def-3 shadow-none;
    }

    & > .icon {
      @apply fg-def-1;

      &.icon-loading {
        color: #0051ff;
      }
    }
  }

  &.icon {
    @apply p-2;
  }

  &:has(> .icon-start):has(> .label) {
    @apply pl-3.5;
  }

  &:has(> .icon-end):has(> .label) {
    @apply pr-3.5;
  }

  & > div.loader {
    @apply w-0 opacity-0;
    @apply top-0 left-0 -mr-2;
    transition: all 0.5s ease;
  }

  &.loading {
    @apply cursor-wait;

    & > div.loader {
      @apply w-4 opacity-100;
      margin-right: revert;
      transition: all 0.5s ease;
    }
  }
}

/* button group */
.button-group .button:first-child {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group .button:first-child {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group .button:last-child {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}
