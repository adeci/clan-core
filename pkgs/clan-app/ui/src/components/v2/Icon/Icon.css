svg.icon {
  @apply fg-def-1;

  &.inverted {
    @apply fg-inv-1;
  }
}
