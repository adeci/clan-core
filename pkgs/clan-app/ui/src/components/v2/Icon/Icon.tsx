import cx from "classnames";
import { Component, JSX, Show, splitProps } from "solid-js";
import ArrowBottom from "@/icons/arrow-bottom.svg";
import ArrowLeft from "@/icons/arrow-left.svg";
import ArrowRight from "@/icons/arrow-right.svg";
import ArrowTop from "@/icons/arrow-top.svg";
import Attention from "@/icons/attention.svg";
import CaretDown from "@/icons/caret-down.svg";
import CaretLeft from "@/icons/caret-left.svg";
import CaretRight from "@/icons/caret-right.svg";
import CaretUp from "@/icons/caret-up.svg";
import Checkmark from "@/icons/checkmark.svg";
import ClanIcon from "@/icons/clan-icon.svg";
import Cursor from "@/icons/cursor.svg";
import Close from "@/icons/close.svg";
import Download from "@/icons/download.svg";
import Edit from "@/icons/edit.svg";
import Expand from "@/icons/expand.svg";
import EyeClose from "@/icons/eye-close.svg";
import EyeOpen from "@/icons/eye-open.svg";
import Filter from "@/icons/filter.svg";
import Flash from "@/icons/flash.svg";
import Folder from "@/icons/folder.svg";
import Grid from "@/icons/grid.svg";
import Info from "@/icons/info.svg";
import List from "@/icons/list.svg";
import Load from "@/icons/load.svg";
import More from "@/icons/more.svg";
import Paperclip from "@/icons/paperclip.svg";
import Plus from "@/icons/plus.svg";
import Reload from "@/icons/reload.svg";
import Report from "@/icons/report.svg";
import Search from "@/icons/search.svg";
import Settings from "@/icons/settings.svg";
import Trash from "@/icons/trash.svg";
import Update from "@/icons/update.svg";
import WarningFilled from "@/icons/warning-filled.svg";
import Modules from "@/icons/modules.svg";
import NewMachine from "@/icons/new-machine.svg";
import AI from "@/icons/ai.svg";
import User from "@/icons/user.svg";
import Heart from "@/icons/heart.svg";
import SearchFilled from "@/icons/search-filled.svg";
import Offline from "@/icons/offline.svg";
import Switch from "@/icons/switch.svg";
import Tag from "@/icons/tag.svg";
import Machine from "@/icons/machine.svg";
import Loader from "@/icons/loader.svg";
import { Dynamic } from "solid-js/web";

const icons = {
  AI,
  ArrowBottom,
  ArrowLeft,
  ArrowRight,
  ArrowTop,
  Attention,
  CaretDown,
  CaretLeft,
  CaretRight,
  CaretUp,
  Checkmark,
  ClanIcon,
  Close,
  Cursor,
  Download,
  Edit,
  Expand,
  EyeClose,
  EyeOpen,
  Filter,
  Flash,
  Folder,
  Grid,
  Heart,
  Info,
  List,
  Load,
  Machine,
  Modules,
  More,
  NewMachine,
  Offline,
  Paperclip,
  Plus,
  Reload,
  Report,
  Search,
  SearchFilled,
  Settings,
  Switch,
  Tag,
  Trash,
  Update,
  User,
  WarningFilled,
};

export type IconVariant = keyof typeof icons;

export interface IconProps extends JSX.SvgSVGAttributes<SVGElement> {
  icon: IconVariant;
  class?: string;
  size?: number | string;
}

const Icon: Component<IconProps> = (props) => {
  const [local, iconProps] = splitProps(props, ["icon", "class"]);

  const IconComponent = () => icons[local.icon];

  return IconComponent() ? (
    <Dynamic
      component={IconComponent()}
      class={cx("icon", local.class)}
      width={iconProps.size || "1em"}
      height={iconProps.size || "1em"}
      viewBox="0 0 48 48"
      // @ts-expect-error: dont know, fix this type nit later
      ref={iconProps.ref}
      {...iconProps}
    />
  ) : undefined;
};

export default Icon;
