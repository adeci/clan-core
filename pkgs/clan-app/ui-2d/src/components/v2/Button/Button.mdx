import { DocsStory, Meta, Canvas } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

# Button

Buttons have a simple hierarchy, `primary` or `secondary`, and two sizes, `default` and `s`.

A `Button` can also have a label with a `startIcon`, an `endIcon` or both:

```tsx
<Button hierarchy="primary">Label</>
<Button hierarchy="secondary" size="s">Label</>
<Button hierarchy="primary" startIcon="Flash">Label</>
<Button hierarchy="primary" size="s" endIcon="Flash">Label</>
<Button hierarchy="primary" startIcon="Flash" endIcon="Flash">Label</>
```

To create a `Button` which is just an icon:

```tsx
<Button icon="Flash"/>
<Button icon="User" size="s"/>
```

<DocsStory of={ButtonStories.Primary} />

<DocsStory of={ButtonStories.Secondary} />

## Ghost Mode

Buttons in ghost mode have all the same states as normal buttons, except for their initial state has no background,
outline or box shadow.

```tsx
<Button hierarchy="primary" ghost={true}>Label</>
```

<DocsStory of={ButtonStories.GhostPrimary} />

<DocsStory of={ButtonStories.GhostSecondary} />
