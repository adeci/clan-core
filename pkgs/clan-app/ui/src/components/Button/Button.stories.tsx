import type { Meta, StoryObj } from "@kachurun/storybook-solid";
import { Button, ButtonProps } from "./Button";
import Icon from "../icon";

const meta: Meta<ButtonProps> = {
  title: "Components/Button",
  component: Button,
};

export default meta;

type Story = StoryObj<ButtonProps>;

const children = "click me";

export const Default: Story = {
  args: {
    children,
  },
};

export const Small: Story = {
  args: {
    ...Default.args,
    size: "s",
  },
};

export const Light: Story = {
  args: {
    ...Default.args,
    variant: "light",
  },
};

export const Ghost: Story = {
  args: {
    ...Default.args,
    variant: "ghost",
  },
};

export const StartIcon: Story = {
  args: {
    ...Default.args,
    startIcon: <Icon size={12} icon="Flash" />,
  },
};

export const EndIcon: Story = {
  args: {
    ...Default.args,
    endIcon: <Icon size={12} icon="Flash" />,
  },
};
