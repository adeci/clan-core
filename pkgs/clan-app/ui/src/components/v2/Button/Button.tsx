import { splitProps, type JSX, createSignal } from "solid-js";
import cx from "classnames";
import { Typography } from "../Typography/Typography";
import { Button as KobalteButton } from "@kobalte/core/button";

import "./Button.css";
import Icon, { IconVariant } from "@/src/components/v2/Icon/Icon";
import { Loader } from "@/src/components/v2/Loader/Loader";

export type Size = "default" | "s";
export type Hierarchy = "primary" | "secondary";

export type Action = () => Promise<void>;

export interface ButtonProps
  extends JSX.ButtonHTMLAttributes<HTMLButtonElement> {
  hierarchy?: Hierarchy;
  size?: Size;
  ghost?: boolean;
  children?: JSX.Element;
  icon?: IconVariant;
  startIcon?: IconVariant;
  endIcon?: IconVariant;
  class?: string;
  onAction?: Action;
}

const iconSizes: Record<Size, string> = {
  default: "1rem",
  s: "0.8125rem",
};

export const Button = (props: ButtonProps) => {
  const [local, other] = splitProps(props, [
    "children",
    "hierarchy",
    "size",
    "ghost",
    "icon",
    "startIcon",
    "endIcon",
    "class",
    "onAction",
  ]);

  const size = local.size || "default";
  const hierarchy = local.hierarchy || "primary";

  const [loading, setLoading] = createSignal(false);

  const onClick = async () => {
    if (!local.onAction) {
      console.error("this should not be possible");
      return;
    }

    setLoading(true);

    try {
      await local.onAction();
    } catch (error) {
      console.error("Error while executing action", error);
    }

    setLoading(false);
  };

  const iconSize = iconSizes[local.size || "default"];

  return (
    <KobalteButton
      class={cx(
        local.class,
        "button", // default button class
        size,
        hierarchy,
        {
          icon: local.icon,
          loading: loading(),
          ghost: local.ghost,
        },
      )}
      onClick={local.onAction ? onClick : undefined}
      {...other}
    >
      <Loader hierarchy={hierarchy} />

      {local.startIcon && (
        <Icon icon={local.startIcon} class="icon-start" size={iconSize} />
      )}

      {local.icon && !local.children && (
        <Icon icon={local.icon} class="icon" size={iconSize} />
      )}

      {local.children && !local.icon && (
        <Typography
          class="label"
          hierarchy="label"
          family="mono"
          size={local.size || "default"}
          color="inherit"
          inverted={local.hierarchy === "secondary"}
          weight="bold"
          tag="span"
        >
          {local.children}
        </Typography>
      )}

      {local.endIcon && (
        <Icon icon={local.endIcon} class="icon-end" size={iconSize} />
      )}
    </KobalteButton>
  );
};
