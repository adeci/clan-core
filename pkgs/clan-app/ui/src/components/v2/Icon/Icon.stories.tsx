import type { Meta, StoryObj } from "@kachurun/storybook-solid";
import { Component, For } from "solid-js";
import Icon, { IconProps, IconVariant } from "./Icon";

const iconVariants: IconVariant[] = [
  "ClanIcon",
  "Checkmark",
  "Paperclip",
  "Expand",
  "Plus",
  "Trash",
  "Folder",
  "CaretRight",
  "CaretLeft",
  "CaretUp",
  "CaretDown",
  "Close",
  "Flash",
  "EyeClose",
  "EyeOpen",
  "Settings",
  "Grid",
  "List",
  "Edit",
  "Load",
  "ArrowRight",
  "ArrowLeft",
  "ArrowTop",
  "ArrowBottom",
  "Info",
  "Update",
  "Reload",
  "Search",
  "Report",
  "Cursor",
  "More",
  "Filter",
  "Download",
  "Attention",
  "User",
  "WarningFilled",
  "Modules",
  "NewMachine",
  "AI",
  "Heart",
  "SearchFilled",
  "Offline",
  "Switch",
  "Tag",
  "Machine",
];

const IconExamples: Component<IconProps> = (props) => (
  <div class="inline-flex flex-wrap gap-2">
    <For each={iconVariants}>{(item) => <Icon {...props} icon={item} />}</For>
  </div>
);

const meta: Meta<IconProps> = {
  title: "Components/Icon",
  component: IconExamples,
};

export default meta;

type Story = StoryObj<IconProps>;

export const Default: Story = {};

export const Large: Story = {
  args: {
    width: "2rem",
    height: "2rem",
  },
};
