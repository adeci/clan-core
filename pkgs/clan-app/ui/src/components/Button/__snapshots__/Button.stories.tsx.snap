// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Button Default smoke-test 1`] = `
<button class="button button--dark button--dark-hover button--dark-focus button--dark-active disabled:bg-secondary-200 disabled:text-secondary-700 disabled:border-secondary-300 button--default">
  <span class="button__icon--start">
    <svg xmlns="http://www.w3.org/2000/svg"
         width="16"
         height="16"
         fill="currentColor"
         viewbox="0 0 48 48"
    >
      <path fill-rule="evenodd"
            d="M24 6.4h3.2v12.8H40v6.4h-3.2v3.2h-3.2V32h-3.2v3.2h-3.2v3.2H24v3.2h-3.2V28.8H8v-6.4h3.2v-3.2h3.2V16h3.2v-3.2h3.2V9.6H24z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </span>
  <span class="text-inherit fnt-clr--inverted fnt-label-default fnt-weight-medium button__label">
    click me
  </span>
</button>
`;

exports[`Components/Button Ghost smoke-test 1`] = `
<button class="button button--ghost-hover button--ghost-focus button--ghost-active button--default">
  <span class="button__icon--start">
    <svg xmlns="http://www.w3.org/2000/svg"
         width="16"
         height="16"
         fill="currentColor"
         viewbox="0 0 48 48"
    >
      <path fill-rule="evenodd"
            d="M24 6.4h3.2v12.8H40v6.4h-3.2v3.2h-3.2V32h-3.2v3.2h-3.2v3.2H24v3.2h-3.2V28.8H8v-6.4h3.2v-3.2h3.2V16h3.2v-3.2h3.2V9.6H24z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </span>
  <span class="text-inherit fnt-label-default fnt-weight-medium button__label">
    click me
  </span>
</button>
`;

exports[`Components/Button Light smoke-test 1`] = `
<button class="button button--light button--light-hover button--light-focus button--light-active button--default">
  <span class="button__icon--start">
    <svg xmlns="http://www.w3.org/2000/svg"
         width="16"
         height="16"
         fill="currentColor"
         viewbox="0 0 48 48"
    >
      <path fill-rule="evenodd"
            d="M24 6.4h3.2v12.8H40v6.4h-3.2v3.2h-3.2V32h-3.2v3.2h-3.2v3.2H24v3.2h-3.2V28.8H8v-6.4h3.2v-3.2h3.2V16h3.2v-3.2h3.2V9.6H24z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </span>
  <span class="text-inherit fnt-label-default fnt-weight-medium button__label">
    click me
  </span>
</button>
`;

exports[`Components/Button Small smoke-test 1`] = `
<button class="button button--dark button--dark-hover button--dark-focus button--dark-active disabled:bg-secondary-200 disabled:text-secondary-700 disabled:border-secondary-300 button button--small">
  <span class="button__icon--start">
    <svg xmlns="http://www.w3.org/2000/svg"
         width="16"
         height="16"
         fill="currentColor"
         viewbox="0 0 48 48"
    >
      <path fill-rule="evenodd"
            d="M24 6.4h3.2v12.8H40v6.4h-3.2v3.2h-3.2V32h-3.2v3.2h-3.2v3.2H24v3.2h-3.2V28.8H8v-6.4h3.2v-3.2h3.2V16h3.2v-3.2h3.2V9.6H24z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </span>
  <span class="text-inherit fnt-clr--inverted fnt-label-s fnt-weight-medium button__label">
    click me
  </span>
</button>
`;
